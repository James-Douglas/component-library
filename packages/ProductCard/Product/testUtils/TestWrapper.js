import React from 'react';
import ProductProvider from '../../Context/ProductProvider';

const ProductTestWrapper = ({
  // eslint-disable-next-line react/prop-types
  children, product, onDetails = () => {}, onSelect = () => {},
}) => (
  <ProductProvider product={product} onDetails={onDetails} onSelect={onSelect}>
    { children }
  </ProductProvider>
);

export default ProductTestWrapper;
