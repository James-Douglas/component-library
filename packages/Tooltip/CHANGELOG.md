# Change Log

All notable changes to this project will be documented in this file.
See [Conventional Commits](https://conventionalcommits.org) for commit guidelines.

## [0.13.2](https://github.com/comparethemarketau/manor-react/compare/@comparethemarketau/manor-tooltip@0.13.1...@comparethemarketau/manor-tooltip@0.13.2) (2020-11-12)

**Note:** Version bump only for package @comparethemarketau/manor-tooltip





## [0.13.1](https://github.com/comparethemarketau/manor-react/compare/@comparethemarketau/manor-tooltip@0.13.0...@comparethemarketau/manor-tooltip@0.13.1) (2020-11-12)

**Note:** Version bump only for package @comparethemarketau/manor-tooltip





# [0.13.0](https://github.com/comparethemarketau/manor-react/compare/@comparethemarketau/manor-tooltip@0.12.0...@comparethemarketau/manor-tooltip@0.13.0) (2020-11-09)


### Features

* **typography:** remove nomargins pro ([#470](https://github.com/comparethemarketau/manor-react/issues/470)) ([ad14285](https://github.com/comparethemarketau/manor-react/commit/ad14285b548116be2306e19bb989113397a11e3f))





# [0.12.0](https://github.com/comparethemarketau/manor-react/compare/@comparethemarketau/manor-tooltip@0.11.0...@comparethemarketau/manor-tooltip@0.12.0) (2020-11-06)


### Features

* **typography:** added help text style ([#469](https://github.com/comparethemarketau/manor-react/issues/469)) ([0827882](https://github.com/comparethemarketau/manor-react/commit/0827882cd6020d5dbada7371faa19660c961af4e))





# [0.11.0](https://github.com/comparethemarketau/manor-react/compare/@comparethemarketau/manor-tooltip@0.10.1...@comparethemarketau/manor-tooltip@0.11.0) (2020-11-04)


### Features

* **tooltips:** fix margin on tooltips with icon ([#464](https://github.com/comparethemarketau/manor-react/issues/464)) ([a0d8773](https://github.com/comparethemarketau/manor-react/commit/a0d877385ccf87ec51685f6832269d200b10b5ed))





## [0.10.1](https://github.com/comparethemarketau/manor-react/compare/@comparethemarketau/manor-tooltip@0.10.0...@comparethemarketau/manor-tooltip@0.10.1) (2020-11-02)


### Bug Fixes

* üêõ syntax fix ([#455](https://github.com/comparethemarketau/manor-react/issues/455)) ([aa1ab46](https://github.com/comparethemarketau/manor-react/commit/aa1ab46df09c07cf3025b0c0e6b64f02e6cfbf85))





# [0.10.0](https://github.com/comparethemarketau/manor-react/compare/@comparethemarketau/manor-tooltip@0.9.4...@comparethemarketau/manor-tooltip@0.10.0) (2020-11-01)


### Features

* **tooltip:** text-anchor tooltip styles & story ([#454](https://github.com/comparethemarketau/manor-react/issues/454)) ([c0efa59](https://github.com/comparethemarketau/manor-react/commit/c0efa59409990791d667e9676507a4e659fa3ead))





## [0.9.4](https://github.com/comparethemarketau/manor-react/compare/@comparethemarketau/manor-tooltip@0.9.3...@comparethemarketau/manor-tooltip@0.9.4) (2020-10-23)


### Bug Fixes

* üêõ tag alignment and tool tip tabbing when inactive ([#441](https://github.com/comparethemarketau/manor-react/issues/441)) ([54f06d4](https://github.com/comparethemarketau/manor-react/commit/54f06d4454c5c66243bcdabeaf316fdd3d6b0211))





## [0.9.3](https://github.com/comparethemarketau/manor-react/compare/@comparethemarketau/manor-tooltip@0.9.2...@comparethemarketau/manor-tooltip@0.9.3) (2020-10-22)

**Note:** Version bump only for package @comparethemarketau/manor-tooltip





## [0.9.2](https://github.com/comparethemarketau/manor-react/compare/@comparethemarketau/manor-tooltip@0.9.1...@comparethemarketau/manor-tooltip@0.9.2) (2020-10-21)

**Note:** Version bump only for package @comparethemarketau/manor-tooltip





## [0.9.1](https://github.com/comparethemarketau/manor-react/compare/@comparethemarketau/manor-tooltip@0.9.0...@comparethemarketau/manor-tooltip@0.9.1) (2020-10-20)

**Note:** Version bump only for package @comparethemarketau/manor-tooltip





# [0.9.0](https://github.com/comparethemarketau/manor-react/compare/@comparethemarketau/manor-tooltip@0.8.6...@comparethemarketau/manor-tooltip@0.9.0) (2020-10-19)


### Features

* **typography:** rewrote typography to no longer use materialUI ([#431](https://github.com/comparethemarketau/manor-react/issues/431)) ([6a7f7db](https://github.com/comparethemarketau/manor-react/commit/6a7f7dbef8f8bea684655f598c2ec9d55b814c34))


### BREAKING CHANGES

* **typography:** props removed: paragraph, gutterBottom. prop changes: noMargins - no defaults to
true, variant - 'button', 'inherit' options removed

* feat(typography): sronly implemented

* feat(typography): address pr feedback

* feat(typography): default display prop to 'block' instead of 'initial'

* feat(typography): update components impacted by typography changes





## [0.8.6](https://github.com/comparethemarketau/manor-react/compare/@comparethemarketau/manor-tooltip@0.8.5...@comparethemarketau/manor-tooltip@0.8.6) (2020-10-02)

**Note:** Version bump only for package @comparethemarketau/manor-tooltip





## [0.8.5](https://github.com/comparethemarketau/manor-react/compare/@comparethemarketau/manor-tooltip@0.8.4...@comparethemarketau/manor-tooltip@0.8.5) (2020-09-29)

**Note:** Version bump only for package @comparethemarketau/manor-tooltip





## [0.8.4](https://github.com/comparethemarketau/manor-react/compare/@comparethemarketau/manor-tooltip@0.8.3...@comparethemarketau/manor-tooltip@0.8.4) (2020-09-29)

**Note:** Version bump only for package @comparethemarketau/manor-tooltip





## [0.8.3](https://github.com/comparethemarketau/manor-react/compare/@comparethemarketau/manor-tooltip@0.8.2...@comparethemarketau/manor-tooltip@0.8.3) (2020-09-28)


### Bug Fixes

* **inputs:** fix styles for inputs ([#409](https://github.com/comparethemarketau/manor-react/issues/409)) ([e866d0e](https://github.com/comparethemarketau/manor-react/commit/e866d0e655d029f6e278183a1e13ea2858ea5070))





## [0.8.2](https://github.com/comparethemarketau/manor-react/compare/@comparethemarketau/manor-tooltip@0.8.0...@comparethemarketau/manor-tooltip@0.8.2) (2020-09-25)

**Note:** Version bump only for package @comparethemarketau/manor-tooltip





# [0.8.0](https://github.com/comparethemarketau/manor-react/compare/@comparethemarketau/manor-tooltip@0.7.0...@comparethemarketau/manor-tooltip@0.8.0) (2020-09-25)


### Features

* **deps:** update all deps again ([#404](https://github.com/comparethemarketau/manor-react/issues/404)) ([7ac9088](https://github.com/comparethemarketau/manor-react/commit/7ac9088c9f042a656b2e8cb2bab615ed44bad88c))





## [0.8.1](https://github.com/comparethemarketau/manor-react/compare/@comparethemarketau/manor-tooltip@0.8.0...@comparethemarketau/manor-tooltip@0.8.1) (2020-09-25)

**Note:** Version bump only for package @comparethemarketau/manor-tooltip





# [0.8.0](https://github.com/comparethemarketau/manor-react/compare/@comparethemarketau/manor-tooltip@0.7.0...@comparethemarketau/manor-tooltip@0.8.0) (2020-09-24)


### Features

* **deps:** update all deps again ([#404](https://github.com/comparethemarketau/manor-react/issues/404)) ([7ac9088](https://github.com/comparethemarketau/manor-react/commit/7ac9088c9f042a656b2e8cb2bab615ed44bad88c))





# [0.7.0](https://github.com/comparethemarketau/manor-react/compare/@comparethemarketau/manor-tooltip@0.6.2...@comparethemarketau/manor-tooltip@0.7.0) (2020-09-24)


### Features

* **rollup:** update rollup to not bundle manor-provider ([#403](https://github.com/comparethemarketau/manor-react/issues/403)) ([0e012d6](https://github.com/comparethemarketau/manor-react/commit/0e012d6fbadcf0ec99857c22e148cacd6265b60a))





## [0.6.2](https://github.com/comparethemarketau/manor-react/compare/@comparethemarketau/manor-tooltip@0.6.1...@comparethemarketau/manor-tooltip@0.6.2) (2020-09-24)


### Bug Fixes

* update manor-provider to latest for Tooltip ([#402](https://github.com/comparethemarketau/manor-react/issues/402)) ([9608553](https://github.com/comparethemarketau/manor-react/commit/9608553319d71f7f11d0b8a8d98cb6b93fadc994))





## [0.6.1](https://github.com/comparethemarketau/manor-react/compare/@comparethemarketau/manor-tooltip@0.6.0...@comparethemarketau/manor-tooltip@0.6.1) (2020-09-23)

**Note:** Version bump only for package @comparethemarketau/manor-tooltip





# [0.6.0](https://github.com/comparethemarketau/manor-react/compare/@comparethemarketau/manor-tooltip@0.5.10...@comparethemarketau/manor-tooltip@0.6.0) (2020-09-22)


### Features

* **provider:** provider context now contains breakpoint & isDesktop ([#398](https://github.com/comparethemarketau/manor-react/issues/398)) ([611b52f](https://github.com/comparethemarketau/manor-react/commit/611b52f8883ca431523b0a8896d51e66ecb76376))





## [0.5.10](https://github.com/comparethemarketau/manor-react/compare/@comparethemarketau/manor-tooltip@0.5.9...@comparethemarketau/manor-tooltip@0.5.10) (2020-09-21)

**Note:** Version bump only for package @comparethemarketau/manor-tooltip





## [0.5.9](https://github.com/comparethemarketau/manor-react/compare/@comparethemarketau/manor-tooltip@0.5.8...@comparethemarketau/manor-tooltip@0.5.9) (2020-09-18)


### Bug Fixes

* **dependencies:** fix more incorrect dependencies ([#392](https://github.com/comparethemarketau/manor-react/issues/392)) ([6fa55a1](https://github.com/comparethemarketau/manor-react/commit/6fa55a11ba89125ccfe61385d9776e4185bff6f3))





## [0.5.8](https://github.com/comparethemarketau/manor-react/compare/@comparethemarketau/manor-tooltip@0.5.7...@comparethemarketau/manor-tooltip@0.5.8) (2020-09-18)

**Note:** Version bump only for package @comparethemarketau/manor-tooltip





## [0.5.7](https://github.com/comparethemarketau/manor-react/compare/@comparethemarketau/manor-tooltip@0.5.6...@comparethemarketau/manor-tooltip@0.5.7) (2020-09-17)

**Note:** Version bump only for package @comparethemarketau/manor-tooltip





## [0.5.6](https://github.com/comparethemarketau/manor-react/compare/@comparethemarketau/manor-tooltip@0.5.5...@comparethemarketau/manor-tooltip@0.5.6) (2020-09-16)

**Note:** Version bump only for package @comparethemarketau/manor-tooltip





## [0.5.5](https://github.com/comparethemarketau/manor-react/compare/@comparethemarketau/manor-tooltip@0.5.4...@comparethemarketau/manor-tooltip@0.5.5) (2020-09-16)

**Note:** Version bump only for package @comparethemarketau/manor-tooltip





## [0.5.4](https://github.com/comparethemarketau/manor-react/compare/@comparethemarketau/manor-tooltip@0.5.3...@comparethemarketau/manor-tooltip@0.5.4) (2020-09-09)

**Note:** Version bump only for package @comparethemarketau/manor-tooltip





## [0.5.3](https://github.com/comparethemarketau/manor-react/compare/@comparethemarketau/manor-tooltip@0.5.2...@comparethemarketau/manor-tooltip@0.5.3) (2020-09-08)

**Note:** Version bump only for package @comparethemarketau/manor-tooltip





## [0.5.2](https://github.com/comparethemarketau/manor-react/compare/@comparethemarketau/manor-tooltip@0.5.1...@comparethemarketau/manor-tooltip@0.5.2) (2020-09-04)

**Note:** Version bump only for package @comparethemarketau/manor-tooltip





## [0.5.1](https://github.com/comparethemarketau/manor-react/compare/@comparethemarketau/manor-tooltip@0.5.0...@comparethemarketau/manor-tooltip@0.5.1) (2020-09-03)

**Note:** Version bump only for package @comparethemarketau/manor-tooltip





# [0.5.0](https://github.com/comparethemarketau/manor-react/compare/@comparethemarketau/manor-tooltip@0.4.0...@comparethemarketau/manor-tooltip@0.5.0) (2020-08-18)


### Features

* **all:** remove manorprovider from components, fix typography props ([#365](https://github.com/comparethemarketau/manor-react/issues/365)) ([9297cf7](https://github.com/comparethemarketau/manor-react/commit/9297cf72e8a7fe8762ec0dadf07d026aa88cbb44))





# [0.4.0](https://github.com/comparethemarketau/manor-react/compare/@comparethemarketau/manor-tooltip@0.3.4...@comparethemarketau/manor-tooltip@0.4.0) (2020-08-17)


* Fix/global styles micro ui (#359) ([bcf8ce9](https://github.com/comparethemarketau/manor-react/commit/bcf8ce92ba170a51113a4022728da22f47a6a768)), closes [#359](https://github.com/comparethemarketau/manor-react/issues/359)


### BREAKING CHANGES

* Global styles have been removed from ManorProvider, as such typography is not
styled by default. Import the Font component once in your application (or macro-ui) and use the
Typography component to render text, and the Link component to render links.

* fix(fonts): fix fonts dependencies

* fix(deps): fix provider dependencies

* fix(feedback): fixes based on PR feedback

Co-authored-by: PSJgit <psjj.dev@gmail.com>





## [0.3.4](https://github.com/comparethemarketau/manor-react/compare/@comparethemarketau/manor-tooltip@0.3.3...@comparethemarketau/manor-tooltip@0.3.4) (2020-08-11)

**Note:** Version bump only for package @comparethemarketau/manor-tooltip





## [0.3.3](https://github.com/comparethemarketau/manor-react/compare/@comparethemarketau/manor-tooltip@0.3.2...@comparethemarketau/manor-tooltip@0.3.3) (2020-08-11)

**Note:** Version bump only for package @comparethemarketau/manor-tooltip





## [0.3.2](https://github.com/comparethemarketau/manor-react/compare/@comparethemarketau/manor-tooltip@0.3.1...@comparethemarketau/manor-tooltip@0.3.2) (2020-08-07)

**Note:** Version bump only for package @comparethemarketau/manor-tooltip





## [0.3.1](https://github.com/comparethemarketau/manor-react/compare/@comparethemarketau/manor-tooltip@0.3.0...@comparethemarketau/manor-tooltip@0.3.1) (2020-07-24)

**Note:** Version bump only for package @comparethemarketau/manor-tooltip





# [0.3.0](https://github.com/comparethemarketau/manor-react/compare/@comparethemarketau/manor-tooltip@0.2.5...@comparethemarketau/manor-tooltip@0.3.0) (2020-07-20)


### Features

* **providers:** remove LayerEvent/Toast providers from manorprovider ([#348](https://github.com/comparethemarketau/manor-react/issues/348)) ([461d724](https://github.com/comparethemarketau/manor-react/commit/461d72498fca1aca9de0056a27d1a3d17a89ea77))


### BREAKING CHANGES

* **providers:** drawers and modals are no longer managed by manor, users will need to use the
zIndex prop (added to both) to do so particularly when multiple layered components are rendered
simultaneously. Additionally, you will need to implement your own keyboard bindings and focus
trapping to ensure conformance to the design system.





## [0.2.5](https://github.com/comparethemarketau/manor-react/compare/@comparethemarketau/manor-tooltip@0.2.4...@comparethemarketau/manor-tooltip@0.2.5) (2020-07-14)


### Bug Fixes

* **toastmanager:** removed portals from ToastManager ([#345](https://github.com/comparethemarketau/manor-react/issues/345)) ([1533377](https://github.com/comparethemarketau/manor-react/commit/1533377910e9cbac266abe24fae1ee42eba4c52f))





## [0.2.4](https://github.com/comparethemarketau/manor-react/compare/@comparethemarketau/manor-tooltip@0.2.3...@comparethemarketau/manor-tooltip@0.2.4) (2020-06-21)

**Note:** Version bump only for package @comparethemarketau/manor-tooltip





## [0.2.3](https://github.com/comparethemarketau/manor-react/compare/@comparethemarketau/manor-tooltip@0.2.2...@comparethemarketau/manor-tooltip@0.2.3) (2020-06-16)

**Note:** Version bump only for package @comparethemarketau/manor-tooltip





## [0.2.2](https://github.com/comparethemarketau/manor-react/compare/@comparethemarketau/manor-tooltip@0.2.1...@comparethemarketau/manor-tooltip@0.2.2) (2020-06-16)

**Note:** Version bump only for package @comparethemarketau/manor-tooltip





## [0.2.1](https://github.com/comparethemarketau/manor-react/compare/@comparethemarketau/manor-tooltip@0.2.0...@comparethemarketau/manor-tooltip@0.2.1) (2020-06-15)

**Note:** Version bump only for package @comparethemarketau/manor-tooltip





# [0.2.0](https://github.com/comparethemarketau/manor-react/compare/@comparethemarketau/manor-tooltip@0.1.0...@comparethemarketau/manor-tooltip@0.2.0) (2020-06-12)


### Features

* **toasts:** manorProvider now injects ToastManager ([#333](https://github.com/comparethemarketau/manor-react/issues/333)) ([8621888](https://github.com/comparethemarketau/manor-react/commit/862188867bbc8258b29fa162f46e5ad5b108f778))


### BREAKING CHANGES

* **toasts:** Added ToastNotification component. Refactored toasts such that ManorProvider
injects the ToastManager. addToasts also now expects a <ToastNotification> instead of an object.

* fix(story): fix typo in toast notification story





# 0.1.0 (2020-06-11)


### Features

* **packages:** migrated to monorepo ([920a45e](https://github.com/comparethemarketau/manor-react/commit/920a45ec4b40a19de32f39f29693cbe1b1f314ae))
