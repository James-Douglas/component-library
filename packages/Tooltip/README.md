# Tooltip

## Installation

`yarn add @comparethemarketau/manor-tooltip`

## Documentation

For technical documentation, see [Storybook](https://services.dev.comparethemarket.cloud/manor/?path=/docs/components-tooltip--tooltip).
Design system documentation can be found here [Bento](https://zeroheight.com/9942937b5/p/703c7d-tooltips/b/97d6bd)
