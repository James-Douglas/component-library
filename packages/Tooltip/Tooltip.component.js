import React, {
  useState, useRef, useEffect, useContext,
} from 'react';
import PropTypes from 'prop-types';
import 'tippy.js/dist/svg-arrow.css';
import 'tippy.js/dist/tippy.css';
import 'tippy.js/animations/scale.css';
import { ManorContext } from '@comparethemarketau/manor-provider';
import { useUnmountEffect, useId } from '@comparethemarketau/manor-hooks';
import { Typography } from '@comparethemarketau/manor-typography';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import { faInfoCircle } from '@fortawesome/pro-light-svg-icons/faInfoCircle';
import {
  StyledTooltipContent,
  StyledTooltipTitle,
  StyledTooltipBody,
  StyledTippy,
  StyledTooltipIcon,
} from './Tooltip.styles';

/**
 * Get the tooltip html content
 * @param title {string} - Optional title for the tooltip
 * @param body {string} - Optional body for the tooltip
 * @returns {*}
 */
export function getContent(title, body, variant) {
  return (
    <StyledTooltipContent>
      {title ? <StyledTooltipTitle variant={variant}><Typography variant="body2" component="span">{title}</Typography></StyledTooltipTitle> : ''}
      <StyledTooltipBody variant={variant}><Typography variant="body2" component="span">{body}</Typography></StyledTooltipBody>
    </StyledTooltipContent>
  );
}

const Tooltip = ({
  id: propsId,
  arrow,
  title,
  body,
  placement,
  variant,
  screenReaderLabelId,
  screenReaderLabel,
  children,
  active,
  className,
}) => {
  const id = useId(propsId);
  const [pinned, setPinned] = useState(false);
  const { isDesktop } = useContext(ManorContext);
  const [tippyInstance, setTippyInstance] = useState(null);
  const [tippyVisible, setTippyVisible] = useState(false);
  const tooltipElement = useRef(null);

  const hideTooltip = () => {
    setTippyVisible(false);
    setPinned(false);
    document.body.removeEventListener('click', hideTooltip);
  };

  const showTooltip = () => {
    if (!tippyVisible) {
      setTippyVisible(true);
    }
  };

  const addOnShowListeners = () => {
    window.addEventListener('scroll', hideTooltip);
  };

  const removeOnShowListeners = () => {
    window.removeEventListener('scroll', hideTooltip);
  };

  useUnmountEffect(() => {
    if (tippyInstance) {
      tippyInstance.destroy();
    }
    document.body.removeEventListener('click', hideTooltip);
    removeOnShowListeners();
  });

  const pinTooltip = (e) => {
    e.preventDefault();
    showTooltip();
    setPinned(true);
  };

  const onTippyShow = () => {
    document.body.click();
    document.body.addEventListener('click', hideTooltip);
    addOnShowListeners();
  };

  const handleKeyPress = (e) => {
    if (e.key === ' ' && e.target === tooltipElement.current) {
      e.preventDefault();
      showTooltip();
    } else if (e.key === 'Escape') {
      e.preventDefault();
      hideTooltip();
    }
  };

  useEffect(() => {
    if (active) {
      showTooltip();
    }
  // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [active]);

  if (!title && !body) {
    return null;
  }

  return (
    <StyledTippy
      content={getContent(title, body, variant)}
      arrow={arrow}
      distance="1rem"
      animation="scale"
      duration={[150, 75]}
      hideOnClick={false}
      trigger="manual"
      onCreate={(instance) => setTippyInstance(instance)}
      onShow={onTippyShow}
      visible={tippyVisible}
      placement={isDesktop ? placement : 'bottom'}
      maxWidth={500}
      delay={125}
      variant={variant}
      className={className}
    >
      {/* eslint-disable-next-line jsx-a11y/no-noninteractive-element-interactions */}
      <StyledTooltipIcon
        id={id}
        tabIndex={active ? '0' : '-1'}
        role="tooltip"
        desktop={isDesktop}
        ref={tooltipElement}
        pinned={pinned}
        tippyVisible={tippyVisible}
        onClick={pinTooltip}
        onKeyDown={handleKeyPress}
        onFocus={showTooltip}
        onBlur={hideTooltip}
        onMouseEnter={showTooltip}
        onMouseLeave={() => {
          if (!pinned) {
            hideTooltip();
          }
        }}
        textAnchor={!!children}
        className={className}
      >
        <>
          <Typography id={screenReaderLabelId} variant="srOnly">{screenReaderLabel}</Typography>
          {
            children || <FontAwesomeIcon icon={faInfoCircle} />
          }
        </>
      </StyledTooltipIcon>
    </StyledTippy>
  );
};

export const tooltipPropTypes = {
  /**
   * An id for the tooltip element
   */
  id: PropTypes.string,
  /**
   * Set the tool tip as active (displayed)
   */
  active: PropTypes.bool,
  /**
   * If the tooltip has an arrow or not
   */
  arrow: PropTypes.bool,
  /**
   * Title for the tooltip
   */
  title: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.array,
    PropTypes.object,
  ]),
  /**
   * Body content for the tooltip
   */
  body: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.array,
    PropTypes.object,
  ]),
  /**
   * Label for screen readers
   */
  screenReaderLabel: PropTypes.string,
  /**
   * id to be applied to the Tooltip screenReaderLabel container for accessibility purposes
   */
  screenReaderLabelId: PropTypes.string,
  /**
   * Placement of the tooltip in relation to it's icon
   */
  placement: PropTypes.oneOf([
    'top', 'bottom', 'left', 'right', 'top-start', 'top-end', 'bottom-start', 'bottom-end', 'left-start', 'left-end',
    'right-start', 'right-end',
  ]),
  /**
   * Light or dark variant (see design system documentation for use cases)
   */
  variant: PropTypes.oneOf(['dark', 'light']),
  /**
   * Classes to be applied to the Tooltip component
   */
  className: PropTypes.string,
  /**
   * Children of the tooltip - replaces the icon
   */
  children: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.node,
    PropTypes.arrayOf(PropTypes.node),
  ]),
};

const defaultProps = {
  id: null,
  arrow: true,
  active: false,
  title: '',
  body: '',
  screenReaderLabel: '',
  screenReaderLabelId: '',
  placement: 'right',
  variant: 'dark',
  className: 'ctm-tooltip',
  children: null,
};

Tooltip.propTypes = tooltipPropTypes;

Tooltip.defaultProps = defaultProps;

export default Tooltip;
