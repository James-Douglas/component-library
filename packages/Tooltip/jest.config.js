const baseConfig = require('../../jest.config.base');

module.exports = {
  ...baseConfig,
  roots: [
    '<rootDir>/packages/Tooltip',
  ],
  collectCoverageFrom: [
    '*.{js,jsx,ts,tsx}',
  ],
  name: 'Tooltip',
  displayName: 'Tooltip',
  rootDir: '../..',
};
