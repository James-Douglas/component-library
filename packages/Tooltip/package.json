{
  "name": "@comparethemarketau/manor-tooltip",
  "version": "0.3.4",
  "description": "Manor Tooltip component",
  "main": "lib/index.cjs.js",
  "author": "Compare The Market",
  "license": "Proprietary",
  "scripts": {
    "build": "rollup -c '../../rollup.config.js'",
    "tdd": "jest --coverage --watchAll",
    "test": "cross-env JEST_SILENT_REPORTER_DOTS=true jest --reporters=jest-silent-reporter --coverage --watchAll",
    "test:ci": "cross-env CI=true jest",
    "prepare": "cp package.json lib/ && cp README.md lib/ && cp CHANGELOG.md lib/"
  },
  "publishConfig": {
    "registry": "https://nexus.comparethemarket.com.au/repository/platform-packages/",
    "access": "public",
    "tag": "latest"
  },
  "dependencies": {
    "@comparethemarketau/manor-hooks": "^0.1.6",
    "@comparethemarketau/manor-typography": "^0.3.0",
    "@fortawesome/fontawesome-svg-core": "^1.2.26",
    "@fortawesome/pro-light-svg-icons": "^5.12.1",
    "@fortawesome/react-fontawesome": "^0.1.8",
    "@tippy.js/react": "^3.1.1",
    "prop-types": "^15.7.2",
    "tippy.js": "^5.1.1"
  },
  "peerDependencies": {
    "react": "^16.11.0",
    "react-dom": "^16.11.0",
    "styled-components": "^5.0.0"
  }
}
