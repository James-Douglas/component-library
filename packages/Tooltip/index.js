import Tooltip, { tooltipPropTypes } from './Tooltip.component';

export { Tooltip, tooltipPropTypes };
