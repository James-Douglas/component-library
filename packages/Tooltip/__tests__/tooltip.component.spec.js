import React from 'react';
import { act } from 'react-dom/test-utils';
import { render, fireEvent } from '../../../testUtils';
import 'jest-styled-components';
import Tooltip, { getContent } from '../Tooltip.component';

describe('getContent()', () => {
  // eslint-disable-next-line react/prop-types
  const ContentContainer = ({ title, body }) => (
    <>
      {getContent(title, body)}
    </>
  );

  it('returns correct content for title only', () => {
    const { container } = render(<ContentContainer title="test title" />);
    expect(container.innerHTML).toMatchSnapshot();
  });

  it('returns correct content for body only', () => {
    const { container } = render(<ContentContainer body="test body" />);
    expect(container.innerHTML).toMatchSnapshot();
  });

  it('returns correct content for title and body', () => {
    const { container } = render(<ContentContainer title="test title" body="test body" />);
    expect(container.innerHTML).toMatchSnapshot();
  });
});

describe('Tooltip', () => {
  beforeEach(() => {
    document.body.innerHTML = '';
  });

  beforeAll(() => {
    global.document.createRange = () => ({
      setStart: () => {},
      setEnd: () => {},
      commonAncestorContainer: {
        nodeName: 'BODY',
        ownerDocument: document,
      },
    });
  });

  it('renders tooltip click', () => {
    const { container } = render(<Tooltip body="test tooltip content" />);
    const tooltipTriggerElement = container.querySelector('[role="tooltip"]');
    fireEvent.click(tooltipTriggerElement);
    const tooltip = document.querySelector('div[data-tippy-root]');
    expect(tooltip).toBeInTheDocument();
  });

  it('renders tooltip hover', () => {
    const { container, getByText } = render(<Tooltip body="test tooltip content" />);
    const tooltipTriggerElement = container.querySelector('[role="tooltip"]');
    fireEvent.mouseOver(tooltipTriggerElement);
    const tooltip = document.querySelector('div[data-tippy-root]');
    expect(tooltip).toHaveStyle('visibility: visible');
    expect(getByText('test tooltip content')).toBeInTheDocument();
    fireEvent.mouseLeave(tooltipTriggerElement);
    expect(tooltip).toHaveStyle('visibility: hidden');
  });

  it('renders tooltip correctly', () => {
    const { container, getByText } = render(<Tooltip body="test tooltip content" />);
    const tooltipTriggerElement = container.querySelector('svg');
    fireEvent.click(tooltipTriggerElement);
    expect(getByText('test tooltip content')).toBeInTheDocument();
  });

  it('dismisses tooltip on escape key', async () => {
    render(<Tooltip body="test tooltip content" />);
    const tooltipTriggerElement = document.body.querySelector('[role="tooltip"]');
    fireEvent.click(tooltipTriggerElement);
    const tooltipElement = document.querySelector('div[data-tippy-root]');
    expect(tooltipElement).toBeInTheDocument();

    fireEvent.keyDown(tooltipTriggerElement, { key: 'Escape', code: 27 });
    expect(document.body.querySelector('div[data-tippy-root]')).not.toBeVisible();
  });

  it('tooltip on click body', async () => {
    render(<Tooltip body="test tooltip content" />);
    const tooltipTriggerElement = document.body.querySelector('[role="tooltip"]');
    fireEvent.click(tooltipTriggerElement);
    const tooltipElement = document.querySelector('div[data-tippy-root]');
    act(() => {
      document.body.click();
    });
    expect(tooltipElement).not.toBeVisible();
  });

  it('dismisses tooltip on click outside', () => {
    render(<Tooltip body="test tooltip content" />);

    const tooltipTriggerElement = document.body.querySelector('[role="tooltip"]');
    fireEvent.click(tooltipTriggerElement);

    const tooltipElement = document.querySelector('div[data-tippy-root]');
    expect(tooltipElement).toBeInTheDocument();

    fireEvent.click(document.body);
    expect(tooltipElement).not.toBeVisible();
  });

  it('dismisses tooltip on blur', () => {
    render(<Tooltip body="test tooltip content" />);

    const tooltipTriggerElement = document.body.querySelector('[role="tooltip"]');
    fireEvent.click(tooltipTriggerElement);

    const tooltipElement = document.querySelector('div[data-tippy-root]');
    expect(tooltipElement).toBeInTheDocument();

    fireEvent.blur(tooltipTriggerElement);
    expect(tooltipElement).not.toBeVisible();
  });
});
