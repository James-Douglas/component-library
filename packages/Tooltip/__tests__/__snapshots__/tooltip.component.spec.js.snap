// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getContent() returns correct content for body only 1`] = `"<div class=\\"sc-AxiKw eyKtTE\\"><div class=\\"sc-AxhUy lirMfo\\"><span class=\\"sc-AxjAm beghoi\\">test body</span></div></div>"`;

exports[`getContent() returns correct content for title and body 1`] = `"<div class=\\"sc-AxiKw eyKtTE\\"><div class=\\"sc-AxhCb yRgYw\\"><span class=\\"sc-AxjAm beghoi\\">test title</span></div><div class=\\"sc-AxhUy lirMfo\\"><span class=\\"sc-AxjAm beghoi\\">test body</span></div></div>"`;

exports[`getContent() returns correct content for title only 1`] = `"<div class=\\"sc-AxiKw eyKtTE\\"><div class=\\"sc-AxhCb yRgYw\\"><span class=\\"sc-AxjAm beghoi\\">test title</span></div><div class=\\"sc-AxhUy lirMfo\\"><span class=\\"sc-AxjAm beghoi\\"></span></div></div>"`;
